### ISSUE TITLE
Sparse indexing: Can't set to itself

### ISSUE THREAD

**Issue by klicperajo** [scipy/scipy#10910]
Since update 1.3 I can't change a fancy indexed sparse matrix in-place directly via `+=`, `-=` etc. Instead, I have to call `.A1` on the right-hand side.

#### Reproducing code example:

```
import numpy as np
import scipy.sparse as sp

a = sp.csr_matrix(np.arange(12).reshape((4,3)))
nz = a.nonzero()
a[nz] = a[nz].A1  # Works
a[nz] = a[nz]  # Error
a[nz] += 1  # Error
```

#### Error message:

```
ValueError                                Traceback (most recent call last)
<ipython-input-1-3f27acc27588> in <module>
      5 nz = a.nonzero()
      6 a[nz] = a[nz].A1  # Works
----> 7 a[nz] = a[nz]  # Error

/path/to/lib/python3.7/site-packages/scipy/sparse/_index.py in __setitem__(self, key, x)
    107             x, _ = np.broadcast_arrays(x, i)
    108             if x.shape != i.shape:
--> 109                 raise ValueError("shape mismatch in assignment")
    110             if x.size == 0:
    111                 return

ValueError: shape mismatch in assignment
```

#### Scipy/Numpy/Python version information:
1.3.0 1.16.4 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)

**Comment by perimosocordiae** [comment]
I believe this is the same issue as gh-10695, which was fixed by gh-10700, which is currently part of the master branch and will be released as part of scipy 1.4.0.

The error no longer appears on master:

```
In [1]: import scipy.sparse as ss

In [2]: a = ss.rand(5, 4, density=0.5, format='csr')

In [3]: a.A
Out[3]: 
array([[0.23017718, 0.        , 0.43185173, 0.80773707],
       [0.2787923 , 0.        , 0.        , 0.87346719],
       [0.31166438, 0.        , 0.60261057, 0.6439705 ],
       [0.        , 0.        , 0.        , 0.        ],
       [0.11345089, 0.        , 0.17125056, 0.        ]])

In [4]: nz = a.nonzero()

In [5]: a[nz]
Out[5]: 
matrix([[0.23017718, 0.43185173, 0.80773707, 0.2787923 , 0.87346719,
         0.31166438, 0.60261057, 0.6439705 , 0.11345089, 0.17125056]])

In [6]: a[nz] = a[nz]

In [7]: import scipy

In [8]: scipy.__version__
Out[8]: '1.4.0.dev0+2d6327e'
```

**Comment by perimosocordiae** [comment]
I just noticed that you call out in-place assignment as well, and I can confirm that's also fixed on master.

Thanks for the well-formed bug report, @klicperajo !
